src/core/lib/event_engine/posix_engine/posix_endpoint.h

```diff
@@ -0,0 +1,637 @@
+      grpc_event_engine::experimental::SliceBuffer* buffer,
+      const grpc_event_engine::experimental::EventEngine::Endpoint::ReadArgs*
+          args) override {
+    impl_->Read(std::move(on_read), buffer, args);
+  }
+
+  void Write(
+      absl::AnyInvocable<void(absl::Status)> on_writable,
+      grpc_event_engine::experimental::SliceBuffer* data,
+      const grpc_event_engine::experimental::EventEngine::Endpoint::WriteArgs*
+          args) override {
+    impl_->Write(std::move(on_writable), data, args);
+  }
+
+  const grpc_event_engine::experimental::EventEngine::ResolvedAddress&
+  GetPeerAddress() const override {
+    return impl_->GetPeerAddress();
+  }
+  const grpc_event_engine::experimental::EventEngine::ResolvedAddress&
+  GetLocalAddress() const override {
+    return impl_->GetLocalAddress();
+  }
+
+  ~PosixEndpoint() override {
+    impl_->MaybeShutdown(absl::InternalError("Endpoint closing"));
+  }
+
+ private:
+  PosixEndpointImpl* impl_;
+};
+
+#else  // GRPC_POSIX_SOCKET_TCP
+
+class PosixEndpoint
+    : public grpc_event_engine::experimental::EventEngine::Endpoint {
+ public:
+  PosixEndpoint() = default;
+
+  void Read(absl::AnyInvocable<void(absl::Status)> /*on_read*/,
+            grpc_event_engine::experimental::SliceBuffer* /*buffer*/,
+            const grpc_event_engine::experimental::EventEngine::Endpoint::
+                ReadArgs* /*args*/) override {
+    GPR_ASSERT(false && "PosixEndpoint::Read not supported on this platform");
+  }
+
+  void Write(absl::AnyInvocable<void(absl::Status)> /*on_writable*/,
+             grpc_event_engine::experimental::SliceBuffer* /*data*/,
+             const grpc_event_engine::experimental::EventEngine::Endpoint::
+                 WriteArgs* /*args*/) override {
+    GPR_ASSERT(false && "PosixEndpoint::Write not supported on this platform");
+  }
+
+  const grpc_event_engine::experimental::EventEngine::ResolvedAddress&
+  GetPeerAddress() const override {
+    GPR_ASSERT(false &&
+               "PosixEndpoint::GetPeerAddress not supported on this platform");
+  }
+  const grpc_event_engine::experimental::EventEngine::ResolvedAddress&
+  GetLocalAddress() const override {
+    GPR_ASSERT(false &&
+               "PosixEndpoint::GetLocalAddress not supported on this platform");
+  }
+
+  ~PosixEndpoint() override = default;
+};
+
+#endif  // GRPC_POSIX_SOCKET_TCP
+
+// Create a PosixEndpoint.
+// A shared_ptr of the EventEngine is passed to the endpoint to ensure that
+// the event engine is alive for the lifetime of the endpoint.
+std::unique_ptr<PosixEndpoint> CreatePosixEndpoint(
+    EventHandle* handle, PosixEngineClosure* on_shutdown,
+    std::shared_ptr<grpc_event_engine::experimental::EventEngine> engine,
+    const grpc_event_engine::experimental::EndpointConfig& config);
+
+}  // namespace posix_engine
+}  // namespace grpc_event_engine
+
+#endif  // GRPC_CORE_LIB_EVENT_ENGINE_POSIX_ENGINE_POSIX_ENDPOINT_H
```

Comment:
missing newline at the end of this file. Did the sanitizer pick that up?

Reply:
Fixed.